tests/integration_tests/functional/test_cpu_template_helper.py

```diff
@@ -0,0 +1,264 @@
+    # should occur and depends on the elapsed time.
+    0x6E0,
+    # MSR_KVM_SYSTEM_TIME_NEW and MSR_KVM_WALL_CLOCK_NEW depend on the elapsed
+    # time.
+    0x4B564D00,
+    0x4B564D01,
+    # MSR_KVM_ASYNC_PF_EN is an asynchronous page fault (APF) control MSR and
+    # is intialized in VM setup process.
+    0x4B564D02,
+    # MSR_KVM_STEAL_TIME indicates CPU steal time filled in by the hypervisor
+    # periodically.
+    0x4B564D03,
+    # MSR_KVM_PV_EOI_EN is PV End Of Interrupt (EOI) MSR and is initialized in
+    # VM setup process.
+    0x4B564D04,
+    # MSR_KVM_ASYNC_PF_INT is an interrupt vector for delivery of 'page ready'
+    # APF events and is initialized just before MSR_KVM_ASYNC_PF_EN.
+    0x4B564D06,
+    # MSR_STAR, MSR_LSTAR, MSR_CSTAR and MSR_SYSCALL_MASK are R/W MSRs that
+    # will be set up by OS to call fast system calls with SYSCALL.
+    0xC0000081,
+    0xC0000082,
+    0xC0000083,
+    0xC0000084,
+    # MSR_AMD64_VIRT_SPEC_CTRL is R/W and can be modified by OS to control
+    # security features for speculative attacks.
+    0xC001011F,
+]
+
+
+def get_guest_msrs(microvm, msr_index_list):
+    """
+    Return the guest MSR in the form of a dictionary where the key is a MSR
+    index and the value is the register value.
+    """
+    msrs_dict = {}
+
+    for index in msr_index_list:
+        if index in MSR_EXCEPTION_LIST:
+            continue
+        rdmsr_cmd = f"rdmsr -0 {index}"
+        code, stdout, stderr = microvm.ssh.execute_command(rdmsr_cmd)
+        assert stderr.read() == "", f"Failed to get MSR for {index=:#x}: {code=}"
+        msrs_dict[index] = int(stdout.read(), 16)
+
+    return msrs_dict
+
+
+@pytest.mark.skipif(
+    PLATFORM != "x86_64",
+    reason=(
+        "`cpuid` and `rdmsr` commands are only available on x86_64. "
+        "System registers are not accessible on aarch64."
+    ),
+)
+def test_cpu_config_dump_vs_actual(
+    test_microvm_with_api_and_msrtools,
+    cpu_template_helper,
+    network_config,
+    tmp_path,
+):
+    """
+    Verify that the dumped CPU config matches the actual CPU config inside guest.
+    """
+    microvm = test_microvm_with_api_and_msrtools
+    microvm.spawn()
+    microvm.basic_config()
+    microvm.ssh_network_config(network_config, "1")
+    vm_config_path = tmp_path / "config.json"
+    save_vm_config(microvm, vm_config_path)
+
+    # Dump CPU config with the helper tool.
+    cpu_config_path = tmp_path / "cpu_config.json"
+    cpu_template_helper.dump(vm_config_path, cpu_config_path)
+    dump_cpu_config = build_cpu_config_dict(cpu_config_path)
+
+    # Retrieve actual CPU config from guest
+    microvm.start()
+    actual_cpu_config = {
+        "cpuid": get_guest_cpuid(microvm, dump_cpu_config["cpuid"].keys()),
+        "msrs": get_guest_msrs(microvm, dump_cpu_config["msrs"].keys()),
+    }
+
+    # Compare actual and dumped CPU config
+    for key in dump_cpu_config["cpuid"]:
```

Comment:
if we iterate over the dumped keys, can we be sure that the actual config doesn't have extra keys that are not dumped?

Reply:
The `cpuid` command result in guests include some leaves/subleaves than the `KVM_GET_SUPPORTED_CPUID`. That's why it only iterates over the dumped keys. Should I switch to do that with some additional exceptions?

Reply:
I changed this to do bidirectionally.
```
    # Compare CPUID between actual and dumped CPU config.
    for (key, actual) in actual_cpu_config["cpuid"].items():
        if (key[0], key[1]) in CPUID_IGNORE_LIST:
            continue
        dump = dump_cpu_config["cpuid"][key]
        if key in CPUID_EXCEPTION_LIST:
            actual &= ~CPUID_EXCEPTION_LIST[key]
            dump &= ~CPUID_EXCEPTION_LIST[key]
        assert actual == dump, (
            f"Mismatched CPUID for leaf={key[0]:#x} subleaf={key[1]:#x} reg={key[2]}:"
            f"{actual=:#034b} vs. {dump=:#034b}"
        )

    # Verify all CPUID on the dumped CPU config are covered in actual one.
    for (key, dump) in dump_cpu_config["cpuid"].items():
        # Skip if the register value is 0.
        if dump == 0:
            continue
        assert actual_cpu_config["cpuid"].get(key) is not None, (
            f"leaf={key[0]:#x} subleaf={key[1]:#x} reg={key[2]} not found in actual CPU config"
        )
```